fpga.chart.py

```diff
@@ -110,45 +96,112 @@ def _connect(self, dsn):
             self.conn = psycopg2.connect(dsn)
             self.conn.autocommit = True
             with self.conn.cursor() as cursor:
-                cursor.execute('CREATE EXTENSION IF NOT EXISTS swarm64da')
+                cursor.execute('SELECT COUNT(*) FROM swarm64da.get_fpga_stats()')
+                self.fpga_count = cursor.fetchone()[0]
 
         except Exception:
             self.conn = None
             return False
 
         return True
 
-    def _parse_fpgainfo(self, cmd, re_string):
-        fpga_info_res = subprocess.check_output(cmd, shell=True)
+
+    def _get_connection(self):
+        if not self.conn:
+            self._connect(self.dsn)
+
+        if self.conn.status != 1:
+            self.conn.close()
+            self.conn = None
+            self._connect(self.dsn)
```

Comment:
Ok, I think we are almost there. Now, please merge `_connect` into `_get_connection` and make `_get_connection` return the actual connection. Otherwise the semantics are too confusing. Likewise, in case of an error on connecting, please let the exception happen for now.